puresnmp/x690/util.py

```diff
@@ -48,8 +63,8 @@ def from_bytes(data: Union[int, bytes]) -> "TypeInfo":
         """
         # pylint: disable=attribute-defined-outside-init
 
-        if isinstance(data, bytes):
-            data = int.from_bytes(data, 'big')
+        if isinstance(data, ( bytes, bytearray)):
```

Comment:
To add to my previous comment about `bytes` and `bytearrays`: I generally try to pay attention to mutability. Introducing `bytearray` makes the value mutable. Essentially using `bytearray` changes the data-type to something which allows dangerous operations. Using `str` instead of `bytearray` would keep the same semantics & behaviour. Unfortunately `str` has different meaning whether you look at it from Python 2 or Python 3. After giving it some thought, I would suggest using [six.binary_type](https://pythonhosted.org/six/#six.binary_type) instead of `bytearray`.

On a bytes-level getting Python2/3 compatibility can be tricky.

Reply:
I'm going to respond here in general to your multiple questions about bytes. I ran into 2 problems with bytes when converting the code to Python 2:

1) bytes(int) behaves very difficult in py2 than py3. However, bytearray(int) behaves the same way, so it seemed the simplest fix.

2) I like your suggestion of overriding bytes() better than using bytearray, so I will get to work making that change.

Reply:
I just googled it, and six.binary_type unfortunately won't fix the problem because of bytes(int) behavior, but creating a bytes function that behaves as desired should do the trick.

Reply:
I'm currently working through the code with your suggestion of overriding bytes, and I will follow it through to completion, but a couple things are bothering me about this approach:

1) Overriding a native type could have unintended consequences and be difficult for other programmers to understand why we chose to do it that way.

2) bytearray has more consistent behavior between py2 and py3

3) bytearray actually performs much better than bytes, see here for example:

http://ze.phyr.us/bytearray/

Reply:
Do you know how to fix it where I can override bytes in py2 without having to rewrite any isinstance ( foo, bytes ) calls?

At the moment it's looking like I'd have to add a bytes_type = bytes and then use bytes_type in every isinstance call.

You wanted less noise, I'm feeling more and more like this approach will not result in less noise, but maybe I'm just not a good enough Python programmer yet :(

Reply:
I'm looking more closely at "six" and I see a few functions that will make my diff cleaner. In particular six.byte2int() and six.indexbytes() will eliminate several uses of bytearray()

Reply:
Okay, I think I found a solution, would you glance at this and give me your feedback?

```python
import six
import sys

def repr2(s):
	# make sure strings always have type prefix ( for the purposes of validating this logic )
	if not isinstance(s, str):
		return repr(s)
	if six.PY2:
		return 'b'+repr(s)
	else:
		return 'u'+repr(s)

try:
	unicode
except NameError:
	unicode = str

# modifications to puresnmp.x690.types.Type class
# so that all classes will inherit the needed behavior
# note that repr2 is only for validation here, we'd use the real repr() in puresnmp
class Type(object):
	def __repr__(self):
		return u'Type({})'.format(repr2(b'foo'))
	def __bytes__(self):
		return b'foo'
	if six.PY2:
		def __unicode__(self):
			return self.__repr__()
		def __str__(self):
			return self.__bytes__()

obj = Type()

def assertEquals ( a, b ):
	assert a == b, ('Assertion Error:\n'
		'   {!r}\n'
		'!= {!r}'.format ( a, b )
	)
assertEquals ( repr2(unicode(obj)), '''u"Type(b'foo')"''' )
b = bytes(obj)
assertEquals ( repr2(b), "b'foo'" )
assertEquals ( repr2(six.byte2int(b)), '102' )
assertEquals ( repr2(six.indexbytes(b,1)), '111' )
```

Reply:
The only complication left is that bytes([40]) does something different on py2.

I feel very strongly against overriding the built-in bytes, so here's what I propose:

1) change those lines from bytes(foo) to to_bytes(foo):

```python
def to_bytes(x):
    return bytes(bytearray(x))
```

2) In order to avoid adding an isinstance() check to to_bytes(), we would need to add one more method to puresnmp.x690.types.Type:

```python
	def __iter__(self):
		return six.iterbytes(self.__bytes__())
```

Would the proposed modifications be acceptable to you?